File: src/main/scala/com/vc/hard/L224.scala
package com.vc.hard

object L224 {
  def calculate(s: String): Int = {
    var res = 0
    var number = 0
    var sign = 1
    val n = s.length
    var i = 0
    import scala.collection.mutable
    val st = new mutable.Stack[Int]()
    while(i < n) {
      val ch = s(i)
      if(Character.isDigit(ch)) {
        number = 0
        while(i < n && Character.isDigit(s(i))) {
          number = 10 * number + (s(i) - '0')
          i += 1
        }
        i -= 1
      }
      else if(ch == '+') {
        res += sign * number
        number = 0
        sign = 1
      }
      else if(ch == '-') {
        res += sign * number
        number = 0
        sign = -1
      }
      else if(ch == '(') {
        st.push(res)
        st.push(sign)
        res = 0
        sign = 1
      }
      else if(ch == ')') {
        res += sign * number
        res *= st.pop
        res += st.pop
        number = 0
      }
      i += 1
    }
    if(number != 0) res += sign * number
    res
  }
}

File: src/main/scala/com/vc/hard/L224.java
package com.vc.hard;

import java.util.*;

class L224 {
    public int calculate(String s) {
        int sum = 0;
        int n = s.length();
        int sign = 1;
        int num = 0;
        int i = 0;
        Stack<Integer> st = new Stack<Integer>();
        while(i < n) {
            char ch = s.charAt(i);
            if(Character.isDigit(ch)) {
                num = num * 10 + (ch - '0');
            }
            else if(ch == '+') {
                sum += sign * num;
                num = 0;
                sign = 1;
            }
            else if(ch == '-') {
                sum += sign * num;
                num = 0;
                sign = -1;
            }
            else if(ch == '(') {
                st.push(sum);
                st.push(sign);
                sign = 1;
                sum = 0;
            }
            else if(ch == ')') {
                int prevSign = st.pop();
                int prevNum = st.pop();
                sum += sign * num;
                sum *= prevSign;
                sum += prevNum;
                num = 0;
            }
            i++;
        }
        if(num != 0) sum += sign * num;
        return sum;
    }
}
